text/0000-unified_coroutines.md

```diff
@@ -0,0 +1,473 @@
+- Feature Name: `unified_coroutines`
+- Start Date: 2019-10-09
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Unify the `Generator` traits with the `Fn*` family of traits. Add a way to pass new arguments upon each resuming of the generator.
+
+# Motivation
+[motivation]: #motivation
+
+The generators/coroutines are extremely powerful concept, but their implementation in Rust is severely limited, and usage requires workarounds in order to achieve useful patterns. The current view of generators is also extremely disconnected from similar concepts, which already exist in the language and the standard library.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider a function. In the olden days, also called a subroutine. This concept is at the core of every programming language, since it is extremely useful. The subroutine has a single entrypoint and a single exit point. Both of these points provide 
+an interface, which is used to transfer data. Upon entering, caller can pass data into the subroutine and upon exiting, the subroutine can return data back to the caller.
+
+Many programming languages(including Rust) also adopted more general concept, callled coroutine. The coroutine , or `Generator` in Rust terms differs from function/subroutine in a single way. It allows the coroutine to `suspend` itself, by storing its state, and `yield`ing control back to the caller, along with data. The caller can then repeatedly pass more data back into the coroutine, and `resume` it. This back-and forth communication is an extremely useful tool for solving a wide class of problems.
+
+Except this is not the truth for Rust's coroutines. The `Generator`, as it  was introduced in order to provide a tool for implementing `async-await`, does not provide this functionality. In order to implement async-await feature, the generators were implemented in their most basic form. And in this form, Generators can't accept arguments, and require invocation of a method, in order to 
+
+These issues severely lessen the usability of the generator feature, and are not difficult to solve.
+
+Current generators take the form of a closure, which contains at least a single yield statement.
+
+```rust
+let name = "World";
+let done = "Done";
+let gen = || {
+    yield "Hello";
+    yield name;
+    return done;
+};
+```
+And are used by calling the `resume` method on the generator.
+```rust
+println!("{}", gen.resume());
```

Comment:
Does this means `gen.resume()` return is display, or you meant "{:?}"?

This applies to almost all `println`s

Reply:
Fix'd